subprojects/solver/solver-z3/src/main/java/hu/bme/mit/theta/solver/z3/Z3Solver.java

```diff
@@ -233,6 +248,68 @@ public void close() {
         z3Context.interrupt();
     }
 
+    private Expr<BoolType> toProofExpr(com.microsoft.z3.Expr<?> expr) {
+        final var args = expr.getArgs();
+        final var lastArg = args[args.length - 1];
+        checkState(lastArg.isApp());
+        final var name = lastArg.getFuncDecl().getName().toString();
+        final var params = lastArg.getArgs();
+        final var paramValues = Arrays.stream(params).map(termTransformer::toExpr).toList();
+        final List<Type> paramTypes = paramValues.stream().map(expr1 -> (Type) expr1.getType()).toList();
+
+        final var funcType = paramTypes.stream().reduce(Bool(), (res, param) -> FuncType.of(param, res));
+        final var decl = Const(name, funcType);
+        Expr<?> func = decl.getRef();
+        for (Expr<?> paramValue : paramValues) {
+            func = UnsafeApp(func, paramValue);
+        }
+        return (Expr<BoolType>) func;
+    }
+
+    /**
+     * This is a best-effort solution, hopefully would support (most) CHCs at least.
+     * Taken from https://github.com/ethereum/solidity/blob/5917fd82b3ca4cab5f817f78b8da8ebe409dd02e/libsmtutil/Z3CHCInterface.cpp#L130
+     * and adapted to the Java API.
+     */
+    @Override
+    public ProofNode getProof() {
+        checkState(status == SolverStatus.UNSAT, "Cannot get proof if status is not UNSAT");
+        com.microsoft.z3.Expr<?> proof = z3Solver.getProof();
+
+        Deque<com.microsoft.z3.Expr<?>> proofStack = new LinkedList<>();
+        proofStack.push(proof.getArgs()[0]);
+
+        Expr<BoolType> root = cast(False(), Bool());
+        final var rootBuilder = new ProofNode.Builder(root);
+
+        Map<Integer, ProofNode.Builder> visited = new LinkedHashMap<>();
+        visited.put(proofStack.peek().getId(), rootBuilder);
+
+        while (!proofStack.isEmpty()) {
+            final var proofNodeExpr = proofStack.pop();
+            if (!visited.containsKey(proofNodeExpr.getId())) {
+                throw new Z3Exception("Node should exist in the graph nodes");
+            }
+            final var proofNode = visited.get(proofNodeExpr.getId());
+
+            if (proofNodeExpr.isApp() && proofNodeExpr.getFuncDecl().getDeclKind() == Z3_decl_kind.Z3_OP_PR_HYPER_RESOLVE) {
+                if (proofNodeExpr.getArgs().length > 0) {
+                    for (int i = 1; i < proofNodeExpr.getArgs().length - 1; ++i) {
+                        final var child = proofNodeExpr.getArgs()[i];
+                        if (!visited.containsKey(child.getId())) {
+                            visited.put(child.getId(), new Builder(toProofExpr(child)));
+                            proofStack.push(child);
+                        }
+                        proofNode.addChild(visited.get(child.getId()));
+                    }
+                }
+            }
+        }
+
+
+        return rootBuilder.build();
+    }
+
```

Comment:
Same as before. Maybe `HornZ3Solver`?

Reply:
Yes, extraction is the way. I'll do this shortly.